Implement test and document BatchServerClass

cmcantalupo opened the pull request:
- Relates to #1962 
- Fixes #1963

cmcantalupo committed:
Add interface implementation and tests for BatchServer class

- Reorder temp file prefix so that "-in" and "-out" are last
- Some white space changes to make more uniform.
- Remove DBusServer.hpp, replace by BatchServer.hpp
- Change BatchServer uses POSIXSignal to shutdown forked process
- Remove message defintions from BatchServer class,
  use only the messages defines in the BatchStatus class
- Add a message to the BatchStatus class for termination

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Move MockSharedMemory into service directory and create symlink

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add mock for BatchStatus object

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

Krebrov001 commented on service/src/BatchServer.cpp:
```diff
@@ -0,0 +1,258 @@
+        , m_is_active(false)
+    {
+        bool is_test = (m_batch_status != nullptr);
+        if (!is_test) {
```
Perhaps invert the if statement?

Krebrov001 commented on service/src/BatchServer.hpp:
```diff
@@ -0,0 +1,167 @@
+            std::string m_server_key;
+            int m_server_pid;
+            bool m_is_active;
+            std::vector<int> m_signal_idx;
```
```cpp
// Each element in the vector is a handle for the request that is pushed onto the stack.
```

Krebrov001 commented on service/src/BatchServer.hpp:
```diff
@@ -0,0 +1,167 @@
+            int m_server_pid;
+            bool m_is_active;
+            std::vector<int> m_signal_idx;
+            std::vector<int> m_control_idx;
```
Would be good to include doxygen comments to describe what exactly is stored in here. And rename them as a vector of **handles**.

Krebrov001 commented on service/src/BatchServer.cpp:
```diff
@@ -0,0 +1,258 @@
+
+        double *buffer = (double *)m_control_shmem->pointer();
+        int buffer_idx = 0;
+        for (const auto &idx : m_control_idx) {
```
I would like to rename `idx` to `handle`. And rename `m_control_idx` to `m_control_handle`.

Krebrov001 commented on service/src/BatchServer.cpp:
```diff
@@ -0,0 +1,258 @@
+        m_pio.read_batch();
+        double *buffer = (double *)m_signal_shmem->pointer();
+        int buffer_idx = 0;
+        for (const auto &idx : m_signal_idx) {
```
Similar for the signals.

Krebrov001 commented on service/src/BatchServer.hpp:
```diff
@@ -0,0 +1,167 @@
+            void check_return(int ret, const std::string &func_name) const;
+
+            int m_client_pid;
+            std::vector<geopm_request_s> m_signal_config;
```
```cpp
const std::vector<geopm_request_s> m_signal_config;
const std::vector<geopm_request_s> m_control_config;
```

Krebrov001 commented on service/src/BatchServer.cpp:
```diff
@@ -0,0 +1,258 @@
+        }
+
+        m_pio.read_batch();
+        double *buffer = (double *)m_signal_shmem->pointer();
```
Can we rename this to `shared_memory_buffer`?

Krebrov001 commented on service/test/MockSharedMemory.hpp:
```diff
@@ -0,0 +1,71 @@
+                    (override));
+
+    protected:
+        std::vector<char> m_buffer;
```
Perhaps we can use `std::array` here instead, since the size of the vector is not changing?

cmcantalupo replied:
Not new code, we can open an issue to change this mock implementation, but I don't see what would motivate that change.

Krebrov001 commented on service/test/BatchServerTest.cpp:
```diff
@@ -0,0 +1,246 @@
+                                                      *m_pio_ptr,
+                                                      m_batch_status,
+                                                      m_signal_shmem,
+                                                      m_control_shmem);
```
Just my style.

cmcantalupo committed:
Change BatchStatus so "-in" and "-out" are at end of temporary file

- Some whitespace clean up to make style consistent

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dannosliwcd commented on service/src/BatchServer.cpp:
```diff
@@ -0,0 +1,250 @@
+                         platform_io(), nullptr, nullptr, nullptr)
+    {
+        // Fork the server when calling real constructor.
+        auto setup = [this]() {this->create_shmem();};
+        auto run = [this]() {this->run_batch();};
+        m_server_pid = fork_with_setup(setup, run);
```
`BatchServerImp::fork_with_setup()` is able to directly call `create_shmem()` and `run_batch`. The `std::function` arguments don't seem necessary.

cmcantalupo replied:
This is to enable testing.  We can substitute other void(void) functions and test that forking them works.

cmcantalupo replied:
A good question.  The answer is that we do not want the lifetime of the shared memory regions to be coupled to the lifetime of the geopmd systemd service process (this should be since server boot).  If we create them after we fork, they should be cleaned up when the forked server ends.

dannosliwcd commented on service/src/BatchServer.cpp:
```diff
@@ -0,0 +1,250 @@
+
+    BatchServerImp::~BatchServerImp()
+    {
+        stop_batch();
```
This calls `BatchStatusImp::send_message` and `BatchStatusImp::receive_message`, which may throw exceptions. Consider wrapping this `stop_batch()` call in a `try...except{log the error}` block, to avoid throwing from a destructor (a possible source of undefined-behavior bugs in cases were a destructor is called due to some other exception, then raises a nested exception from itself).

dannosliwcd replied:
Also, is it possible to hang on destruction if it's waiting for a message that never comes?

cmcantalupo replied:
Good idea!

dannosliwcd commented on service/src/BatchServer.cpp:
```diff
@@ -0,0 +1,250 @@
+        int forked_pid = fork();
+        check_return(forked_pid, "fork(2)");
+        if (forked_pid == 0) {
+            check_return(close(pipe_fd[0]), "close(2)");
```
Looks like the reason for this pipe is to ensure that the parent process does not proceed until the child process has created the shared memory. Why not just create the shared memory before forking?

cmcantalupo replied:
A good question.  The answer is that we do not want the lifetime of the shared memory regions to be coupled to the lifetime of the geopmd systemd service process (this should be since server boot).  If we create them after we fork, they should be cleaned up when the forked server ends.

cmcantalupo replied:
I have reset the shared memory pointers and am now calling _Exit().  This should clean up any stale keys that could get left around in certain error conditions, but does not call the destructor for all of the IOGroups registered in PlatformIO :)

dannosliwcd reviewed (commented):
Partially done. Left off at BatchServerTest.cpp

cmcantalupo reviewed (commented):
Thanks for the feedback Daniel

cmcantalupo commented on service/src/BatchServer.cpp:
```diff
@@ -0,0 +1,250 @@
+
+    void BatchServerImp::stop_batch(void)
+    {
+        if (m_is_active) {
+            m_batch_status->send_message(M_MESSAGE_QUIT);
+            m_batch_status->receive_message(M_MESSAGE_CONTINUE);
+            m_is_active = false;
```
This message is going the wrong way.  The caller of this method is the parent process that forked the batch server (i.e. the primary geopmd process).  It doesn't make sense for geopmd to send a message to the client process to "quit".  This should be probably be resolved with a signal sent by geopmd and a signal handler on the batch server process.

cmcantalupo reviewed (commented):
The BatchServerImp::stop_batch() is not working in integration tests.  The reason is because of the comment above.

cmcantalupo commented on service/src/BatchServer.cpp:
```diff
@@ -0,0 +1,250 @@
+    void BatchServerImp::run_batch(void)
+    {
+        push_requests();
+        m_batch_status->send_message(M_MESSAGE_CONTINUE);
```
This messsage is extra, no continue should be sent at this point

dannosliwcd reviewed (commented):
Finished the pass where I left off. No additional comments from the remaining files.

cmcantalupo committed:
Add mock for POSIXSignal class

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

Krebrov001 committed:
Adding new tests for `BatchServer`

 - `BatchServerTest.run_batch_read_empty`
 - `BatchServerTest.run_batch_write_empty`
 - `BatchServerTest.create_shmem`
 - `BatchServerTest.fork_with_setup`

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

cmcantalupo committed:
Some clean up of unit tests for BatchServerTest

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Avoid most situations where a thrown exception would hang a client request

- Use _Exit to avoid calling destructors on static objects
- We do want to reset the shared memory pointers that were
  created after the fork.
- We do not want to delete the singleton objects (like PlatformIO)
  when the forked server exits.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented on service/src/BatchServer.cpp:
```diff
@@ -0,0 +1,338 @@
+                stop_batch();
+            }
+            catch (const Exception &ex) {
+                std::cerr << "Warning: <geopm> BatchServerImp::~BatchServerImp(): Exception thrown in destructor: " <<  ex.what();
+            }
+            catch (...) {
+                std::cerr << "Warning: <geopm> BatchServerImp::~BatchServerImp(): Non-GEOPM exception thrown in destructor";
```
add `\n` to end of messages

cmcantalupo commented on service/test/BatchServerTest.cpp:
```diff
@@ -0,0 +1,434 @@
+
+    std::function<void(void)> setup = [&counter](void)
+    {
+        if(counter == 0u) {
```
```suggestion
        if (counter == 0u) {
```

cmcantalupo commented on service/test/BatchServerTest.cpp:
```diff
@@ -0,0 +1,434 @@
+        PROT_READ  | PROT_WRITE,     // int prot
+        MAP_SHARED | MAP_ANONYMOUS,  // int flags
+        -1,                          // int fd
+        0                            // off_t offset
+    );
```
```suggestion
        0);                          // off_t offset
```

Krebrov001 commented on service/test/BatchServerTest.cpp:
```diff
@@ -74,20 +73,19 @@ class BatchServerTest : public ::testing::Test
         std::vector<geopm_request_s> m_signal_config;
         std::vector<geopm_request_s> m_control_config;
         std::shared_ptr<BatchServerImp> m_batch_server;
+        std::shared_ptr<BatchServerImp> m_batch_server_signals;
+        std::shared_ptr<BatchServerImp> m_batch_server_controls;
+        std::shared_ptr<BatchServerImp> m_batch_server_pid;
```
What is the pro of having three different `BatchServerImp` objects, instead of having only one, whose definition changes based on the test fixture? It seems that only one of the three objects is used at any given time.

cmcantalupo replied:
This enables us to easily write more test fixtures that use any of these objects, and avoids issues around calling the destructor if we overwrite a the m_batch_server to set it to a new value.  It also allows the TearDown function to be responsible for more of the clean up after each test, thus reducing lines of code in each test fixture.

Krebrov001 commented on service/test/BatchServerTest.cpp:
```diff
@@ -285,34 +264,28 @@ TEST_F(BatchServerTest, run_batch_read_empty)
 
-    m_batch_server->run_batch();
-
-    EXPECT_EQ(result.size(), m_signal_shmem->size());  // both should be == 0
```
Is that the reason why we cannot check the size, because `nullptr` is passed in as the argument for the `m_signal_shmem` in the `BatchServerImp::BatchServerImp()`?

Krebrov001 commented on service/test/BatchServerTest.cpp:
```diff
@@ -242,23 +237,11 @@ TEST_F(BatchServerTest, run_batch_read)
-            .WillOnce(Return(idx))
-            .RetiresOnSaturation();
-        ++idx;
-    }
```
I notice that we avoid pushing the signals by not having that for loop at all, since our `m_signal_config` is still initialized with the proper values. And the empty vector is passed into the constructor instead.
```cpp
m_signal_config = {geopm_request_s {1, 0, "CPU_FREQUENCY"},
                       geopm_request_s {2, 1, "TEMPERATURE"}};
```

Krebrov001 commented on service/test/BatchServerTest.cpp:
```diff
@@ -395,77 +352,78 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+    int forked_pid = m_batch_server_pid->fork_with_setup(setup, run);
     waitpid(forked_pid, NULL, 0);
 
     ASSERT_EQ(2u, counter);
```
```suggestion
    EXPECT_EQ(2u, counter);
```

Krebrov001 commented on service/test/BatchServerTest.cpp:
```diff
@@ -395,77 +352,78 @@ TEST_F(BatchServerTest, run_batch_write_empty)
+            ++counter;
+        }
+        this->TearDown();
+        testing::Mock::AllowLeak(m_signal_shmem.get());
+        testing::Mock::AllowLeak(m_control_shmem.get());
+        testing::Mock::AllowLeak(m_pio_ptr.get());
+        testing::Mock::AllowLeak(m_batch_status.get());
+        testing::Mock::AllowLeak(m_posix_signal.get());
```
Can we do something like
```cpp
this->m_signal_shmem.reset();
```

cmcantalupo replied:
If there are any copies of the shared pointer elsewhere, the reset() will not be sufficient.  In this case we know that the _Exit() call will not clean up these mocks, and so we can use the gmock feature.

cmcantalupo replied:
m_signal_shmem and m_control_shmem should *not* be leaked since they are managed by the forked process explicitly.  These two lines should be removed.

Krebrov001 commented on service/src/BatchServer.cpp:
```diff
@@ -78,7 +79,20 @@ namespace geopm
+            }
+            catch(...) {
+                if (m_is_client_waiting) {
+                    std::cerr << "Warning: <geopm>: " << __FILE__ << ":" << __LINE__
+                              << " Batch server was terminated while client was waiting: sending client quit message\n";
```
Minor note, it seems that the `BatchServer` may be invoked by the `geopmd` as just a server process, without any connection to the command line ... perhaps we could use the `syslog()` instead for writing error messages? I have used it before.

cmcantalupo replied:
This detail is managed by dasbus.

Krebrov001 commented on service/test/BatchServerTest.cpp:
```diff
@@ -285,34 +264,28 @@ TEST_F(BatchServerTest, run_batch_read_empty)
-    m_batch_server->run_batch();
-
-    EXPECT_EQ(result.size(), m_signal_shmem->size());  // both should be == 0
+    m_batch_server_controls->run_batch();
```
Hmm ... do we want this one to be `run_batch()` or `event_loop()`? It seems that we aren't handling any theoretical thrown exceptions in this particular unit test. I assume that these exceptions can occur upon receipt of any signal other than `SIGTERM`? Maybe it can be used to test the failure condition?

cmcantalupo replied:
Would be great to get better testing of cases where mock objects in the unit tests are made to throw exceptions when they are called.

Krebrov001 commented on service/src/BatchServer.cpp:
```diff
@@ -144,42 +162,85 @@ namespace geopm
+            }
+            catch (const Exception &ex) {
+                // If we were not interupted SIGTERM rethrow
+                if (ex.err_value() != EINTR ||
```
Why are we not setting the `in_message = BatchStatus::M_MESSAGE_TERMINATE;` in this situation?

cmcantalupo replied:
I have refactored this code a bit in an attempt to clarify what is going on for the reader.  Let me know if it makes more sense now.

Krebrov001 reviewed (commented):
Thank you Chris for providing me the opportunity to review your pull request.

Krebrov001 commented:
In general, I would like to add the doxygen `@throws` to all methods which may throw an exception, either directly or indirectly.
https://www.doxygen.nl/manual/commands.html

dannosliwcd commented on service/src/BatchServer.cpp:
```diff
@@ -78,7 +79,20 @@ namespace geopm
+                this->run_batch();
+            }
+            catch(...) {
+                if (m_is_client_waiting) {
```
This looks redundant with the check inside `run_batch`, so one of these two loggers can probably be removed. This one _should_ be dead code in its current state since the other one clears `m_is_client_waiting` before rethrowing. 

I think removing either one would be fine. I'd probably favor removing the deeper-in-the-callstack log-and-rethrows since it should result in fewer instances of manually bubbling errors up a callstack.

cmcantalupo commented on service/src/BatchServer.cpp:
```diff
@@ -0,0 +1,373 @@
+            check_return(write(pipe_fd[1], &msg, 1), "write(2)");
+            check_return(close(pipe_fd[1]), "close(2)");
+            run();
+            exit(0);
```
use _exit() and reset the shared memory pointers.

cmcantalupo committed:
Change assert to expect to increase verbosity when test fails

Co-authored-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

cmcantalupo commented on service/src/BatchServer.cpp:
```diff
@@ -177,17 +174,15 @@ namespace geopm
-                }
+        try {
+            m_batch_status->send_message(out_message);
+	    m_is_client_waiting = false;
```
whitespace!
